Ansible Deployment for Windows (#646)
signalfx/splunk-otel-collector: deployments/ansible/roles/collector/README.md

collect metrics, traces and logs from Linux machines and send data to [Splunk 
Observability Cloud](https://www.splunk.com/en_us/observability.html). 

> ## Windows

Reviewer: It is confusing that a Windows header exists, but not a Linux -- made me initially think this page was just for Windows. If we add for Windows we should add for Linux as well and section header generic.

Author: Yes, I will add Linux as well.

Currently, the following Windows versions are supported:

> Ansible requires PowerShell 3.0 or newer and atleast .NET4.0 to be installed on Windows host.

Reviewer: ```suggestion
- Ansible requires PowerShell 3.0 or newer and at least .NET 4.0 to be installed on Windows host.
```

> A WinRM listner should be created and activeted. 

Reviewer: Not a Windows/Ansible expert -- not sure what this is or how to ensure available

Author: For detailed setup of Ansible in windows I have added Ansible docs reference link.

> A WinRM listner should be created and activeted. 

Reviewer: ```suggestion
- A WinRM listener should be created and activated. 
```

collect metrics, traces and logs from Linux machines and send data to [Splunk 
Observability Cloud](https://www.splunk.com/en_us/observability.html). 

## Linux
Currently, the following Linux distributions and versions are supported:

- Amazon Linux
- RedHat
- Debian: buster, stretch
- Ubuntu: 16.04, 18.04, 20.04

## Windows
Currently, the following Windows versions are supported:
- Ansible requires PowerShell 3.0 or newer and at least .NET 4.0 to be installed on Windows host.

> A WinRM listner should be created and activated. 

Reviewer 2: ```suggestion
A WinRM listener should be created and activated. 
```

> For setting up Windows Host Refer : [Ansible Docs](https://docs.ansible.com/ansible/latest/user_guide/windows_setup.html)

Reviewer 2: ```suggestion
For setting up Windows Host refer [Ansible Docs](https://docs.ansible.com/ansible/latest/user_guide/windows_setup.html)
```

- Windows Server 2012 64-bit
- Windows Server 2016 64-bit

> - Windows Server 2019 64-bit

Reviewer 2: Can we move this list higher right under "Currently, the following Windows versions are supported:" and separate it from the list of Windows requirements

```yaml
- name: Install Splunk OpenTelemetry Connector
  hosts: all

>   #For Windows "become: yes" will raise error.
>   #"The Powershell family is incompatible with the sudo become plugin" Remove "become: yes" tag to run on Windows

Reviewer: ```suggestion
  # For Windows "become: yes" will raise error.
  # "The Powershell family is incompatible with the sudo become plugin" Remove "become: yes" tag to run on Windows
```

```yaml
- name: Install Splunk OpenTelemetry Connector
  hosts: all

>   become_user: root

Reviewer 2: Same comment here. Why changed? AFAIK `become_user: root` is default anyway, but it doesn't work if there is no `become: yes`.

- `splunk_otel_collector_config`: Splunk OTel Collector config YAML file. Can be set to 
  `/etc/otel/collector/gateway_config.yaml` to install the collector in gateway
  mode. (**default:** `/etc/otel/collector/agent_config.yaml` on Linux, 
  **default:** `%ProgramData%\Splunk\OpenTelemetry Collector\agent_config.yaml` on Windows)

- `splunk_otel_collector_config_source`: Source path to a Splunk OTel Collector config YAML 
  file on your control host that will be uploaded and set in place of
  `splunk_otel_collector_config` in remote hosts. Can be used to submit a custom collector 
  config, e.g. `./custom_collector_config.yaml`. (**default:** `""` 
  meaning that nothing will be copied and existing `splunk_otel_collector_config` will be used)

>   

Reviewer 2: I think I commented on this already. Please discard this change

>   

Reviewer 2: Please discard this change, looks unrelated

- `splunk_bundle_dir`: The path to the [Smart Agent bundle directory](
  https://github.com/signalfx/splunk-otel-collector/blob/main/internal/extension/smartagentextension/README.md).
  The default path is provided by the collector package. If the specified path
  is changed from the default value, the path should be an existing directory
  on the node. The `SPLUNK_BUNDLE_DIR` environment variable will be set to
  this value for the collector service. (**default:**
  `/usr/lib/splunk-otel-collector/agent-bundle` on Linux, **default:** 
  `%ProgramFiles%\Splunk\OpenTelemetry Collector\agent-bundle` on Windows)

- `splunk_collectd_dir`: The path to the collectd config directory for the
  Smart Agent bundle. The default path is provided by the collector package.
  If the specified path is changed from the default value, the path should be
  an existing directory on the node. The `SPLUNK_COLLECTD_DIR` environment
  variable will be set to this value for the collector service.
  (**default:** `/usr/lib/splunk-otel-collector/agent-bundle` on Linux, 
  **default:** `%ProgramFiles%\Splunk\OpenTelemetry Collector\agent-bundle\run\collectd` on Windows)

> 

Reviewer 2: nit: extra space

- `splunk_otel_collector_config`: Splunk OTel Collector config YAML file. Can be set to 
  `/etc/otel/collector/gateway_config.yaml` to install the collector in gateway
  mode. (**default:** `/etc/otel/collector/agent_config.yaml`)
  For Windows this is the destination path of the collector config file on the node. 
  The SPLUNK_CONFIG environment variable will be set with this value for the collector service.

>   (**default:** `%ProgramData%\Splunk\OpenTelemetry Collector\agent_config.yaml`)

Reviewer 2: Why the definition is different for Windows? It should have the same meaning. So we can just change the default value to the following:
**default:** `/etc/otel/collector/agent_config.yaml` on Linux, `%ProgramData%\Splunk\OpenTelemetry Collector\agent_config.yaml` on Windows

- `splunk_otel_collector_config_source`: Source path to a Splunk OTel Collector config YAML 
  file on your control host that will be uploaded and set in place of
  `splunk_otel_collector_config` in remote hosts. Can be used to submit a custom collector 
  config, e.g. `./custom_collector_config.yaml`. (**default:** `""` meaning 
  that nothing will be copied and existing `splunk_otel_collector_config` will be used)
  For Windows this file will be copied to the $splunk_otel_collector_config path on the node.

>   (**default:** `%ProgramFiles%\Splunk\OpenTelemetry Collector\agent_config.yaml`)

Reviewer 2: Same Q as above, another definition for Windows is confusing. The parameter should have the same meaning and behaviour across platforms

  
- `splunk_bundle_dir`: The path to the [Smart Agent bundle directory](
  https://github.com/signalfx/splunk-otel-collector/blob/main/internal/extension/smartagentextension/README.md).
  The default path is provided by the collector package. If the specified path
  is changed from the default value, the path should be an existing directory
  on the node. The `SPLUNK_BUNDLE_DIR` environment variable will be set to

>   this value for the collector service. For Linux (**default:**
>   `/usr/lib/splunk-otel-collector/agent-bundle`). For Windows (**default:** 
>   `%ProgramFiles%\Splunk\OpenTelemetry Collector\agent-bundle`)

Reviewer 2: ```suggestion
  this value for the collector service. (**default:**
  `/usr/lib/splunk-otel-collector/agent-bundle` on Linux,  
  `%ProgramFiles%\Splunk\OpenTelemetry Collector\agent-bundle` on Windows)
```

- `splunk_collectd_dir`: The path to the collectd config directory for the
  Smart Agent bundle. The default path is provided by the collector package.
  If the specified path is changed from the default value, the path should be
  an existing directory on the node. The `SPLUNK_COLLECTD_DIR` environment

>   variable will be set to this value for the collector service. For Linux
>   (**default:** `/usr/lib/splunk-otel-collector/agent-bundle`). For Windows(**default:** 
>   `%ProgramFiles%\Splunk\OpenTelemetry Collector\agent-bundle\run\collectd`)

Reviewer 2: ```suggestion
  variable will be set to this value for the collector service. (**default:** 
  `/usr/lib/splunk-otel-collector/agent-bundle` on Linux, 
  `%ProgramFiles%\Splunk\OpenTelemetry Collector\agent-bundle\run\collectd` on Windows)
```

  config, e.g. `./custom_collector_config.yaml`. (**default:** `""` meaning 
  that nothing will be copied and existing `splunk_otel_collector_config` will be used)

- `collector_path_win`: Default path of Splunk-otel-collector in windows 
  (**default:** `C:\\Program Files\Splunk\OpenTelemetry Collector\`)

- `splunk_bundle_dir` & : The path to the [Smart Agent bundle directory](
  https://github.com/signalfx/splunk-otel-collector/blob/main/internal/extension/smartagentextension/README.md).
  The default path is provided by the collector package. If the specified path
  is changed from the default value, the path should be an existing directory
  on the node. The `SPLUNK_BUNDLE_DIR` environment variable will be set to
  this value for the collector service.  (**default:**

>   `/usr/lib/splunk-otel-collector/agent-bundle` For Windows `${collector_path_win}\agent-bundle`)

Reviewer: This Linux versus Windows thing is hard to read -- perhaps all these options can be turned into a table

Parameter (default) | OS | Description

Where OS is Linux|Windows|All

Reviewer: Another alternative is to have different sections per OS -- this would result in duplicated information, but perhaps fine as it make it more clear

Author: Sure will update in that format.

  stretch, and `4.1.1` for other distros`)

- `splunk_fluentd_config`: Path to the fluentd config file on the remote host.
  (**default:** `/etc/otel/collector/fluentd/fluent.conf` on Linux, 
  **default:** `%SYSTEMDRIVE%\opt\td-agent\etc\td-agent\td-agent.conf` on Windows)

- `splunk_fluentd_config_source`: Source path to a fluentd config file on your 
  control host that will be uploaded and set in place of `splunk_fluentd_config` on
  remote hosts. Can be used to submit a custom fluentd config,
  e.g. `./custom_fluentd_config.conf`. (**default:** `""`)

>   meaning that nothing will be copied and existing `splunk_fluentd_config` will be used)

Reviewer 2: Same comment, parentheses mismatch

>   meaning that nothing will be copied and existing `splunk_fluentd_config` will be used)

Reviewer 2: Please discard this change. Try to keep you diffs always related to the change you are working on

- `splunk_fluentd_config`: Path to the fluentd config file on the remote host.
  (**default:** `/etc/otel/collector/fluentd/fluent.conf`)

> - `splunk_fluentd_config_source`: Source path to a fluentd config file on your 

Reviewer 3: This is still an active variable? Why is the description being removed?

Author: Added back.

- `splunk_fluentd_config`: Path to the fluentd config file on the remote host.
  (**default:** `/etc/otel/collector/fluentd/fluent.conf`)

## Windows Specific Role Variables

- `collector_path_win`: Default path of Splunk-otel-collector in windows 

>   (**default:** `C:\\Program Files\Splunk\OpenTelemetry Collector\`)

Reviewer 4: The MSI installation directory is not currently configurable and installs to `%PROGRAMFILES%\Splunk\OpenTelemetry Collector`.

> - `splunk_bundle_dir_win` & : The path to the [Smart Agent bundle directory](

Reviewer 4: Is the `&` needed here?

  https://github.com/signalfx/splunk-otel-collector/blob/main/internal/extension/smartagentextension/README.md).
  The default path is provided by the collector package. If the specified path
  is changed from the default value, the path should be an existing directory
  on the node. The `SPLUNK_BUNDLE_DIR` environment variable will be set to
  this value for the collector service.  (**default:** `${collector_path_win}\agent-bundle`)

- `splunk_collectd_dir_win`: The path to the collectd config directory for the
  Smart Agent bundle. The default path is provided by the collector package.
  If the specified path is changed from the default value, the path should be
  an existing directory on the node. The `SPLUNK_COLLECTD_DIR` environment
  variable will be set to this value for the collector service.
  (**default:** `${splunk_bundle_dir_win}\run\collectd`)

> - `splunk_fluentd_config_source`: Source path to a fluentd config file on your 

Reviewer 4: Should this be moved out of the windows section?

  control host that will be uploaded and set in place of `splunk_fluentd_config` on
  remote hosts. Can be used to submit a custom fluentd config,
  e.g. `./custom_fluentd_config.conf`. (**default:** `""` meaning 
  that nothing will be copied and existing `splunk_fluentd_config` will be used)

- `collector_config_source_win`: Source path to the collector config YAML file. This file will 

>   be copied to the $collector_config_dest path on the node. See the source attribute of the file 

Reviewer 4: The `See the source attribute of the file resource for supported value types.` statement is not applicable to ansible and should be removed.

  stretch, and `4.1.1` for other distros`)

- `splunk_fluentd_config`: Path to the fluentd config file on the remote host.

>   For Linux(**default:** `/etc/otel/collector/fluentd/fluent.conf`). 
>   For Windows (**default:** `%SYSTEMDRIVE%\opt\td-agent\etc\td-agent\td-agent.conf`)

Reviewer 2: ```suggestion
  (**default:** `/etc/otel/collector/fluentd/fluent.conf` on Linux, 
  `%SYSTEMDRIVE%\opt\td-agent\etc\td-agent\td-agent.conf` on Windows)
```

- `splunk_fluentd_config_source`: Source path to a fluentd config file on your 
  control host that will be uploaded and set in place of `splunk_fluentd_config` on
  remote hosts. Can be used to submit a custom fluentd config,
  e.g. `./custom_fluentd_config.conf`. For Linux (**default:** `""` meaning 
  that nothing will be copied and existing `splunk_fluentd_config` will be used)
  For Windows this file will be copied to the $splunk_fluentd_config path on the node.

>   (**default:** `%ProgramFiles%\Splunk\OpenTelemetry Collector\fluentd\td-agent.conf`)

Reviewer 2: Same comment
